### ISSUE TITLE
Unable to build generated typescript code

### ISSUE THREAD

**Issue by jabalsad** [awslabs/smithy-typescript#492]
Hi,

When running `npm build` or `yarn build` within the root of the `typescript-codegen` directory, we get a few compile time errors:

1. This error repeats once for every operation in the specification.
```
src/commands/UpdateTransactionGeneratorCommand.ts:70:13 - error TS2339: Property 'logger' does not exist on type 'LedgerServiceClientResolvedConfig'.

70     const { logger } = configuration;
               ~~~~~~
```

2. Interface error

```
src/ts/commands/CreateAccountCommand.ts:67:45 - error TS2345: Argument of type 'LedgerServiceClientResolvedConfig' is not assignable to parameter of type 'SerdeContext'.
  Property 'endpoint' is missing in type 'SmithyConfiguration<HttpHandlerOptions> & Required<ClientDefaults>' but required in type 'SerdeContext'.

67     this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));

```

Any advice would be appreciated to correct these issues.

**Comment by adamthom-amzn** [comment]
Which release of smithy-typescript-codegen are you using? We run a CI job in the JS SDK repository to make sure that a very simple model can generate and build: https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/generic-client-test-codegen/build.gradle.kts

I don't have enough information to be certain yet, but my guess is either you're not including the AWS codegen library (`software.amazon.smithy.typescript:smithy-aws-typescript-codegen`), there's some sort of version mismatch, or the generator version is from before we attempted to ensure that non-AWS clients were functional.

**Comment by jabalsad** [comment]
Hi and thanks for your feedback on the matter!

I was simply following similar instructions to the readme on this repo, and the little bit I've learnt from using the Go code generators.

I have not included the `smithy-aws-typescript-codegen` dependency, as the readme only lists the `smithy-typescript-codegen` dependency. Specifically, the include line in our build.gradle.kts looks like this:

```
    implementation("software.amazon.smithy.typescript:smithy-typescript-codegen:0.8.0")
```

The smithy-typescript-codegen dependency I had to build and publish to the local maven repo myself (it could not resolve any public dependencies). I did this by checking out `smithy-typescript` and running `pushd smithy-typescript-codegen && ./gradlew build` which created the 0.8.0 artificat in my local maven repo. The commit I have is from January 11th 2022 (so fairly recent).

To be clear: generating the code works, but compiling the generating code does not (due to the error supplied in the first message of this issue).

Here is a complete copy of our build.gradle.kts:

```
// This example builds the model and places it in the JAR.

plugins {
    id("software.amazon.smithy").version("0.6.0")
}

repositories {
    mavenLocal()
    mavenCentral()
}

buildscript {
    dependencies {
        classpath("software.amazon.smithy:smithy-openapi:1.11.0")
        classpath("software.amazon.smithy.go:smithy-go-codegen:0.1.0")
        // classpath("software.amazon.smithy:smithy-aws-apigateway-openapi:1.11.0")
        // The openapi plugin configured in the smithy-build.json example below
        // uses the restJson1 protocol defined in the aws-traits package. This
        // additional dependency must added to use that protocol.
        classpath("software.amazon.smithy:smithy-aws-traits:1.11.0")
    }
}

dependencies {
    implementation("software.amazon.smithy:smithy-model:[1.0, 2.0[")
    implementation("software.amazon.smithy:smithy-aws-traits:[1.0, 2.0[")
    implementation("software.amazon.smithy.go:smithy-go-codegen:0.1.0")
    implementation("software.amazon.smithy.typescript:smithy-typescript-codegen:0.8.0")
}

configure<software.amazon.smithy.gradle.SmithyExtension> {
    // Uncomment this to use a custom projection when building the JAR.
    // projection = "foo"
}

// Uncomment to disable creating a JAR.
tasks["jar"].enabled = false
```

As well as our smithy-build.json

```json
{
  "version": "1.0",
  "outputDirectory": "build",
  "plugins": {
    "go-codegen": {
      "service": "abc.myservice#MyService",
      "module": "github.com/jabalsad/my-service/src/go/ledger"
    },
    "typescript-codegen": {
      "service": "abc.myservice#MyService",
      "package": "my-service-client",
      "packageVersion": "0.0.1",
      "packageJson": {
        "license": "Apache-2.0"
      }
    },
    "openapi": {
      "service": "abc.myservice#MyService",
      "protocol": "aws.protocols#restJson1",
      "useIntegerType": true,
      "jsonAdd": {
        "/servers": [
          {
            "url": "https://test.com/v1"
          }
        ]
      }
    }
  }
}
```

**Comment by jabalsad** [comment]
I tested out building using the `smithy-aws-typescript-codegen` dependency instead, and this time the code generation does not even work successfully:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':smithyBuildJar'.
> 'software.amazon.smithy.model.shapes.ShapeId software.amazon.smithy.model.shapes.OperationShape.getInputShape()'
```

With verbose logging enabled:
```
Caused by: java.lang.NoSuchMethodError: 'software.amazon.smithy.model.shapes.ShapeId software.amazon.smithy.model.shapes.OperationShape.getInputShape()'
        at software.amazon.smithy.model.validation.validators.OperationValidator.validateOperationNameAmbiguity(OperationValidator.java:152)
```

**Comment by adamthom-amzn** [comment]
Yeah, that happens when there's a version mismatch between smithy dependencies in your project. `smithy-aws-typescript-codegen` uses APIs from smithy 1.15, and your other dependencies are not that recent.

You should be able to fix this with a `buildscript` dependency on smithy-cli 1.15.0. This example is in groovy format, not kotlin, but hopefully illustrative:

```
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
       classpath("software.amazon.smithy:smithy-cli:1.15.0")
    }
}
```

**Comment by jabalsad** [comment]
To clarify - should I be using `smithy-typescript-codegen:0.8.0` or `smithy-aws-typescript-codegen:0.8.0` ?

**Comment by adamthom-amzn** [comment]
> To clarify - should I be using `smithy-typescript-codegen:0.8.0` or `smithy-aws-typescript-codegen:0.8.0` ?

I'd like to see what happens when you use `smithy-aws-typescript-codegen`. A lot of packages specify both, but since `smithy-aws-typescript-codegen` depends on `smithy-typescript-codegen` you shouldn't need it.

**Comment by jabalsad** [comment]
We upgraded our dependencies and added `smithy-cli` as a dependency. FYI I find it interesting that depending on `smithy-cli:1.16.3` breaks again but `smithy-cli:1.15` works.

**Comment by jabalsad** [comment]
OK - depending on `smithy-aws-typescript-codegen:0.8.0` has resolved the `tsc` build problem and the TS code all compiles successfully after a `yarn install`. I'm still not sure what caused the `smithy-typescript-codegen` version to fail, since I pulled in all the latest changes and built it again to test.

Thanks for helping me work through it, though I recommend upgrading your readme to reflect the recommended steps.

**Comment by jabalsad** [comment]
Lastly (before we can close this issue :-) ); can you elaborate perhaps on why the `1.15` smithy tools seem to work but the `1.16.x` ones do not? Is this something that is expected to be fixed in the future if we ever have a need to upgrade?

**Comment by adamthom-amzn** [comment]
> Lastly (before we can close this issue :-) ); can you elaborate perhaps on why the `1.15` smithy tools seem to work but the `1.16.x` ones do not? Is this something that is expected to be fixed in the future if we ever have a need to upgrade?

Smithy 1.16 will be supported in the 0.9 code generators; see https://github.com/aws/aws-sdk-js-v3/pull/3215 and https://github.com/awslabs/smithy-typescript/pull/499

> though I recommend upgrading your readme to reflect the recommended steps.

We need a mix of this and some work to remove the necessity of relying on the `aws` generator.

**Comment by jabalsad** [comment]
Sounds good, thank you!

**Comment by jabalsad** [comment]
@adamthom-amzn

I'd like to piggy-back on this issue, because I haven't quite been able to build the typescript package. I found a bug in the typescript codegen:

The generated package.json has a `types` field that points to `dist-types/index.d.ts` but the directory tree that is contained in `source-projection/source/typescript-codegen` has all the types located under a `src` tree in `dist-types/src/index.d.ts`.

This results in being unable to import this package as a dependency from npm because no types are found.

**Comment by adamthom-amzn** [comment]
I can't reproduce this with our generic client that we use for CI tests (https://github.com/aws/aws-sdk-js-v3/tree/main/codegen/generic-client-test-codegen). `dist-types` isn't generated by codegen, but created by `tsc` during `build:types`. What does that target look like in your generated package.json, and what does your generated `tsconfig.types.json` look like?

**Comment by jabalsad** [comment]
Smithy generates the following tsconfig.types.json file:
```
{
  "extends": "./tsconfig",
  "compilerOptions": {
    "removeComments": false,
    "declaration": true,
    "declarationDir": "dist-types",
    "emitDeclarationOnly": true
  },
  "exclude": ["test/**/*", "dist-types/**/*"]
}

```

And the `types` field in the generated `package.json` points to `dist-types/index.d.ts`

**Comment by adamthom-amzn** [comment]
And I'm assuming `build:types` is `"tsc -p tsconfig.types.json",` in your package.json?

if your `rootDir` is `src` in `tsconfig.json` I don't know why there would be a `src` folder in your `dist-types` folder. There is not for me when I build locally.

**Comment by jabalsad** [comment]
Yes. The package.json and tsconfig files are all untouched (I'm using them as is from the generated smithy output).

**Comment by jabalsad** [comment]
It appears I've tracked it down to a subtle difference in the command that is used to invoke the build (`yarn` vs `npm`).

If I run `yarn build` in the root of the `typescript-codegen` directory, I end up with `dist-*` directories that contain the `src` dir. However, if I instead run `npm run build`, these `src` directories don't appear.

**Comment by jamesls** [comment]
FWIW I ran into this as well and needed to use the `smithy-aws-typescript-codegen` instead of the `smithy-typescript-codegen`.

The logger error appears to be because the aws codegen version adds additional properties that aren't in the `smithy-typescript-codegen` version, one of which is `logger`.


<details>
  <summary><strong>diff smithy vs. smithy-aws</strong></summary>

```
--- /tmp/smithy-only/src/MyServiceClient.ts	2022-07-06 14:23:58.000000000 -0400
+++ /tmp/smithy-aws/src/MyServiceClient.ts	2022-07-06 16:37:19.000000000 -0400

export interface ClientDefaults
  extends Partial<__SmithyResolvedConfiguration<__HttpHandlerOptions>> {

...

+   * Value for how many times a request will be made at most in case of retry.
+   */
+  maxAttempts?: number | __Provider<number>;
+
+  /**
+   * Specifies which retry algorithm to use.
+   */
+  retryMode?: string | __Provider<string>;
+
+  /**
+   * Optional logger for logging debug/info/warn/error.
+   */
+  logger?: __Logger;
+
+  /**
+   * Enables IPv6/IPv4 dualstack endpoint.
+   */
+  useDualstackEndpoint?: boolean | __Provider<boolean>;
+
+  /**
+   * Enables FIPS compatible endpoints.
+   */
+  useFipsEndpoint?: boolean | __Provider<boolean>;
+
+  /**
+   * Unique service identifier.
+   * @internal
+   */
+  serviceId?: string;
+
+  /**
+   * Fetch related hostname, signing name or signing region with given region.
+   * @internal
+   */
+  regionInfoProvider?: RegionInfoProvider;
+
+  /**
+   * The provider populating default tracking information to be sent with `user-agent`, `x-amz-user-agent` header
+   * @internal
+   */
+  defaultUserAgentProvider?: Provider<__UserAgent>;
}
```

</details>


The other error I get, which might be new from when the original issue was created, is around endpoints:

```
[build:cjs] src/commands/MyCommand.ts(47,45): error TS2345: Argument of type 'MyServiceClientResolvedConfig' is not assignable to parameter of type 'SerdeContext'.
[build:cjs]   Property 'endpoint' is missing in type 'SmithyConfiguration<HttpHandlerOptions> & Required<ClientDefaults>' but required in type 'SerdeContext'.
```

Which looks like it's coming from `EndpointsResolvedConfig` when comparing the diff of `smithy-typescript` vs. `smithy-aws-typescript`:


<details>
  <summary><strong>diff smithy vs. smithy-aws</strong></summary>

```
 type MyServiceClientResolvedConfigType = __SmithyResolvedConfiguration<__HttpHandlerOptions>
   & Required<ClientDefaults>
+  & RegionResolvedConfig
+  & EndpointsResolvedConfig
+  & RetryResolvedConfig
+  & HostHeaderResolvedConfig
+  & UserAgentResolvedConfig
```

</details>

Using `smithy-aws-typescript-codegen` does result in having to provide an
unnecessary `region` param when creating a client (I'm providing my own
`endpoint` param), but otherwise seems to work.  

NOTE: I also had to pull in #561 to get everything working for real, otherwise `npm install` fails with the error from #560.

**Comment by instantdelay** [comment]
It looks like the original problem here has finally been fixed in this PR: https://github.com/awslabs/smithy-typescript/pull/693

From what I can tell, `smithy-aws-typescript-codegen` extends `smithy-typescript-codegen` with various plugins. One of those plugins was responsible for adding `logger` to the generated runtime configuration. Some code in this package was incorrectly assuming that the runtime config would always contain `logger` even though nothing here was adding it, so builds with `smithy-aws-typescript-codegen` would succeed but generic builds with this package would fail.  It's now added by `AddClientRuntimeConfig.java`

**Comment by gosar** [comment]
Thanks @instantdelay. Yes, https://github.com/awslabs/smithy-typescript/pull/693 resolves the main issue here. There were a couple of other topics in this issue, but I think they are addressed. If there is still another issue, please open a new specific issue.
